cpp/src/c_api/ecg.cpp

```diff
@@ -0,0 +1,154 @@
+              size_t ensemble_size,
+              size_t max_level,
+              double threshold,
+              double resolution,
+              bool do_expensive_check)
+    : abstract_functor(),
+      handle_(*reinterpret_cast<cugraph::c_api::cugraph_resource_handle_t const*>(handle)->handle_),
+      rng_state_(reinterpret_cast<cugraph::c_api::cugraph_rng_state_t*>(rng_state)),
+      graph_(reinterpret_cast<cugraph::c_api::cugraph_graph_t*>(graph)),
+      max_level_(max_level),
+      threshold_(threshold),
+      resolution_(resolution),
+      do_expensive_check_(do_expensive_check)
+  {
+  }
+
+  template <typename vertex_t,
+            typename edge_t,
+            typename weight_t,
+            typename edge_type_type_t,
+            bool store_transposed,
+            bool multi_gpu>
+  void operator()()
+  {
+    if constexpr (!cugraph::is_candidate<vertex_t, edge_t, weight_t>::value) {
+      unsupported();
+    } else {
+      // ecg expects store_transposed == false
+      if constexpr (store_transposed) {
+        error_code_ = cugraph::c_api::
+          transpose_storage<vertex_t, edge_t, weight_t, store_transposed, multi_gpu>(
+            handle_, graph_, error_.get());
+        if (error_code_ != CUGRAPH_SUCCESS) return;
+      }
+
+      auto graph =
+        reinterpret_cast<cugraph::graph_t<vertex_t, edge_t, false, multi_gpu>*>(graph_->graph_);
+
+      auto graph_view = graph->view();
+
+      auto edge_weights = reinterpret_cast<
+        cugraph::edge_property_t<cugraph::graph_view_t<vertex_t, edge_t, false, multi_gpu>,
+                                 weight_t>*>(graph_->edge_weights_);
+
+      auto number_map = reinterpret_cast<rmm::device_uvector<vertex_t>*>(graph_->number_map_);
+
+      rmm::device_uvector<vertex_t> clusters(0, handle_.get_stream());
+
+      weight_t modularity;
+
+      std::tie(clusters, std::ignore, modularity) =
+        cugraph::ecg(handle_,
+                     rng_state_->rng_state_,
+                     graph_view,
+                     (edge_weights != nullptr)
+                       ? std::make_optional(edge_weights->view())
+                       : std::make_optional(cugraph::c_api::create_constant_edge_property(
+                                              handle_, graph_view, weight_t{1})
+                                              .view()),
+                     static_cast<weight_t>(min_weight_),
+                     ensemble_size_,
+                     max_level_,
+                     static_cast<weight_t>(threshold_),
+                     static_cast<weight_t>(resolution_));
+
+      rmm::device_uvector<vertex_t> vertices(graph_view.local_vertex_partition_range_size(),
+                                             handle_.get_stream());
+      raft::copy(vertices.data(), number_map->data(), vertices.size(), handle_.get_stream());
+
+      result_ = new cugraph::c_api::cugraph_hierarchical_clustering_result_t{
+        modularity,
```

Comment:
python ECG returns as results vertices with their corresponding `partition/cluster` therefore `modularity` will be new and should be exposed to the PLC and python API